`hasData` service fixes and improvements

Dosant opened the pull request:
## Summary

**In this pr:** 

- Rename `hasUserIndexPattern` -> `hasDataView` because I believe that naming discrepancy was the culprit of the introduced bug where hasUserIndexPattern [was refactored](https://github.com/elastic/kibana/pull/135322) to check if there are any indices instead of data views. 
- Fix server side `hasDataView` logic to check data views (fix https://github.com/elastic/kibana/issues/137567). I noticed that in 8.4 we no longer have any data views created by default in the cloud, making the logic redundant. So instead of fixing it, I cleaned it up. I still left the `hasUserDataView` abstraction if we need to hardcode something there again.  I think this clean-up should have been part of https://github.com/elastic/kibana/issues/127869
- I added `logs-enterprise_search.audit-default` as a known default data stream. this fixes https://github.com/elastic/kibana/issues/134918. Longer-term enterprise search should stop creating streams by default and we should find a way to prevent adding new stuff by default as we will constantly be breaking on-boarding flow otherwise. 
- On this client side, I changed `hasData` service to return `true` instead of `false` in case of errors: https://github.com/elastic/kibana/blob/24de6906b56186c2ed7f53dc664468bc810fc3b6/src/plugins/data_views/public/services/has_data.ts#L109 . I believe that falling back to `true` and assuming that data exists is safer, as otherwise Kibana could stuck in an onboarding state in case there is some error. It is also more work to handle manually on consumer side to let each usage decide.
- Added more tests. Especially around search fallback to checking es data.

@Dosant to create a separate issue to improve: 

- After introducing new `hadData` API we didn't remove old data view `has_user_index_pattern`, I think we should remove it and rewrite all the functional test to check new API 
- We need to add functional tests when we fallback to search for checking es data (https://github.com/elastic/kibana/issues/132478) I believe we don't have such tests
- ~When we check remote clusters~ (https://github.com/elastic/kibana/pull/137824#discussion_r935852171), we don't filter known data sources as we do when we check local. Probably we should. Also, probably, we should see if we could do this when we fallback to search. https://github.com/elastic/kibana/blob/fff39b0ee0ee22d734b79ebdfd75e11477c6f941/src/plugins/data_views/public/services/has_data.ts#L139

### Checklist

Delete any items that are not applicable to this PR.

- [x] [Unit or functional tests](https://www.elastic.co/guide/en/kibana/master/development-tests.html) were updated or added to match the most common scenarios

### Risk Matrix

Delete this section if it is not applicable to this PR.

Before closing this PR, invite QA, stakeholders, and other developers to identify risks that should be tested prior to the change/feature release.

When forming the risk matrix, consider some of the following examples and how they may potentially impact the change:

| Risk                      | Probability | Severity | Mitigation/Notes        |
|---------------------------|-------------|----------|-------------------------|
| Double check permission use cases https://github.com/elastic/kibana/issues/132478  | Low | High |  |



### For maintainers

- [ ] This was checked for breaking API changes and was [labeled appropriately](https://www.elastic.co/guide/en/kibana/master/contributing.html#kibana-release-notes-process)


<!--ONMERGE {"backportTargets":["8.4"]} ONMERGE-->

majagrubic commented on src/plugins/data_views/common/constants.ts:
```diff
@@ -22,15 +22,16 @@ export const RUNTIME_FIELD_TYPES = [
+ * Used to optimize on-boarding experience to determine if the instance has some user created data views or data indices/streams by filtering data sources
+ * that are created by default by elastic in ese.
+ * We should somehow prevent creating initial data for the users without their explicit action
+ * instead of relying on these hardcoded assets
```
100% agree, but until we do that, can we return the old data view names?

Dosant replied:
Do you mean to keep the older logic from `hasUserDataView`? As I understand, we don't need that because https://github.com/elastic/kibana/issues/82851 was fixed, and we should have to remove it in https://github.com/elastic/kibana/issues/127869. 

Do you think there is a reason to keep it? If yes, then I'd also revert https://github.com/elastic/kibana/issues/127869

majagrubic replied:
Just the old data view names, `logs-*`, `logs-elastic_agent` etc

majagrubic commented on src/plugins/data_views/public/services/has_data.ts:
```diff
@@ -156,22 +158,33 @@ export class HasData {
+      .catch((e) => {
+        // eslint-disable-next-line no-console
+        console.warn(`hasDataViews failed with error, assuming there are data views`, e);
+        return true;
```
🙌

majagrubic replied:
Also fyi, most of the UI is already catching this in the same way.

Dosant replied:
It is more often `false` in consuming code. those catches are redundant, but if we'd remove them from services, then probably should indeed change most of those to `true` 

<img width="958" alt="Screenshot 2022-08-02 at 16 02 32" src="https://user-images.githubusercontent.com/7784120/182393829-3340e479-7b1f-4c22-8b51-658044ce4a04.png">

mattkime commented on src/plugins/data_views/public/services/has_data.test.ts:
```diff
@@ -54,11 +54,83 @@ describe('when calling hasData service', () => {
+    const spy = jest.spyOn(http, 'get').mockImplementation((path: any) =>
+      Promise.resolve({
+        aliases: [],
+        data_streams: path.includes('*:*') // TODO: should we also filter default assets for remote cluster check?
```
>>TODO: should we also filter default assets for remote cluster check?

I think that would be a good idea

Dosant replied:
Sounds good 👍 will remove the comment

majagrubic reviewed (approved):
Synced with @Dosant offline. It seems we purposely don't want to be ignoring `logs-*` data views.

Dosant commented:
> Synced with @Dosant offline. It seems we purposely don't want to be ignoring `logs-*` data views.

👍 

I'll just elaborate in case someone will wonder: 

In 7.14 we've added the code that ignored `logs-*` `metrics-*` data views if they were backed up only by some specific known data sources. We've added this because when you created a deployment in the cloud, fleet automatically created those data views without any prior user action and this broke the onboarding flow. Fleet fixed this in 8.1, as they stopped creating those data views automatically: https://github.com/elastic/kibana/issues/82851. At that point, as I understand, that code where we ignore data views became redundant.

Then I believe after some shuffling around with `hasData` service a bug was introduced in that code is causing https://github.com/elastic/kibana/issues/137567. Instead of fixing the code, we are removing it. 
This should be safe because when you create a new 8.4 deployment, there is no automatically created `logs-*` data view anymore, so onboarding kicks in

kertal reviewed (approved):
Approving by testing in Discover. Locally, with a fresh Kibana without this fix, I've created a `logs-` prefixed index in Elasticsearch. This caused the empty Discover page. Then I applied the fix of this PR and the Discover page was no longer empty 👍 thx for the fix

mattkime reviewed (approved):
Code changes look good and work well, thanks for addressing this!

kibanamachine commented:
## 💚 All backports created successfully

| Status | Branch | Result |
|:------:|:------:|:------|
|✅|8.4|[<img src="https://img.shields.io/github/pulls/detail/state/elastic/kibana/138043">](https://github.com/elastic/kibana/pull/138043)|

Note: Successful backport PRs will be merged automatically after passing CI.

### Questions ?
Please refer to the [Backport tool documentation](https://github.com/sqren/backport)

<!--- Backport version: 8.9.0 -->
